system/Publisher/ContentReplacer.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Publisher;
+
+use RuntimeException;
+use stdClass;
+
+class ContentReplacer
+{
+    /**
+     * Replace content
+     *
+     * @param array $replaces [search => replace]
+     */
+    public function replace(string $content, array $replaces): string
+    {
+        return strtr($content, $replaces);
+    }
+
+    /**
+     * Add text
+     *
+     * @param string $text    Text to add.
+     * @param string $pattern Regexp search pattern.
+     * @param string $replace Regexp replacement including text to add.
+     *
+     * @return stdClass {updated: bool, content: string}
+     */
+    private function add(string $content, string $text, string $pattern, string $replace): stdClass
+    {
+        $result = new stdClass();
+
+        $return = preg_match('/' . preg_quote($text, '/') . '/u', $content);
+
+        if ($return === false) {
+            // Regexp error.
+            throw new RuntimeException(preg_last_error_msg());
```

Comment:
preg_last_error_msg() requires PHP8.

Reply:
😕 You're right. Looks like we can use https://www.php.net/manual/en/function.preg-last-error.php with the constants, but it's definitely more handling. Maybe just make the error message something like "Regex failed with error #__"?

Reply:
Yes. Done.

Reply:
Looks good!